src/riak_kv_ts_svc.erl

```diff
@@ -355,147 +355,27 @@ sub_tslistkeysreq(Mod, DDL, #tslistkeysreq{table = Table,
 sub_tscoveragereq(Mod, _DDL, #tscoveragereq{table = Table,
                                             query = Q},
                   State) ->
-    {select, SQL} = decode_query(Q, undefined),
-    case compile(Mod, SQL) of
-        {error, #rpberrorresp{} = Error} ->
-            {reply, Error, State};
-        {error, _} ->
+    Client = {riak_client, [node(), undefined]},
+    %% all we need from decode_query is to compile the query,
+    %% but also to check permissions
+    case decode_query(Q, undefined) of
+        {_QryType, {ok, SQL}} ->
+            case riak_kv_ts_api:compile_to_per_quantum_queries(Mod, SQL) of
+                {ok, Compiled} ->
+                    Bucket = riak_kv_ts_util:table_to_bucket(Table),
+                    {reply,
+                     {tscoverageresp, riak_kv_ts_util:sql_to_cover(Client, Compiled, Bucket, [])},
+                     State};
+                {error, Reason} ->
+                    {reply, make_rpberrresp(
+                      ?E_BAD_QUERY, flat_format("Failed to compile query: ~p", [Reason])), State}
+            end;
+        {error, Reason} ->
             {reply, make_rpberrresp(
-                      ?E_BAD_QUERY, "Failed to compile query"),
-             State};
-        Queries ->
-            %% SQL is a list of queries (1 per quantum)
-            Bucket = riak_kv_ts_util:table_to_bucket(Table),
-            Client = {riak_client, [node(), undefined]},
-            convert_cover_list(sql_to_cover(Client, Queries, Bucket, []), State)
+                      ?E_BAD_QUERY, flat_format("Failed to parse query: ~p", [Reason])),
```

Comment:
A proper error retval matching is in order. I did this for the main tsqueryreq case; time to fix it here, too.